sql/semisync_master.cc

```diff
@@ -922,6 +915,34 @@ int Repl_semi_sync_master::commit_trx(const char *trx_wait_binlog_name,
         }
       }
 
+      /* In between the binlogging of this transaction and this wait, it is
+       * possible that our entry in Active_tranx was removed (i.e. if
+       * semi-sync was switched off and on). It is also possible that the
+       * event was already sent to a replica; however, we don't know if
+       * semi-sync was on or off at that time, so an ACK may never come. So
+       * skip the wait. Note that rpl_semi_sync_master_request_acks was
+       * already incremented in report_binlog_update(), so to keep
+       * rpl_semi_sync_master_yes/no_tx consistent with it, we check for a
+       * semi-sync restart _after_ checking the reply state.
+       */
+      if (unlikely(!m_active_tranxs->is_thd_waiter(thd, trx_wait_binlog_name,
+                                                   trx_wait_binlog_pos)))
+      {
+        sql_print_information(
+            "Skipping semi-sync wait for transaction at pos %s, %lu. This is "
+            "likely because semi-sync turned off and on during the lifetime "
+            "of this transaction.",
+            trx_wait_binlog_name, trx_wait_binlog_pos);
+
```

Comment:
I don't think this sql_print_information() is needed (just make it a DBUG_PRINT() instead).
If semi-sync is disabled during the transaction, it is expected that it will not wait for ack. And a message like this triggered by a specific timing between re-enabling and commit processing is more confusing than helpful.